modules/thermo-cycler/thermo-cycler-arduino/gcode.cpp

```diff
@@ -0,0 +1,141 @@
+#include "gcode.h"
+
+Gcode::Gcode() {
+}
+
+void Gcode::_strip_serial_buffer() {
+  for(uint8_t i=0;i<SERIAL_BUFFER_STRING.length();i++) {
+    for (uint8_t c=0;c<sizeof(CHARACTERS_TO_STRIP);c++) {
+      if(SERIAL_BUFFER_STRING[i] == CHARACTERS_TO_STRIP[c]) {
+        SERIAL_BUFFER_STRING.remove(i, 1);
+        i--;
+        break;
+      }
+    }
+  }
+}
+
+bool Gcode::pop_command() {
+  _strip_serial_buffer();
+  GCODE_BUFFER_STRING += SERIAL_BUFFER_STRING;
+  SERIAL_BUFFER_STRING = "";
+  code = GCODE_NO_CODE;
+  while (GCODE_BUFFER_STRING.length()) {
+    for (uint8_t i=GCODE_GET_LID_STATUS; i<=TOTAL_GCODE_COMMAND_CODES; i++) {
+      if (GCODE_BUFFER_STRING.substring(0, COMMAND_CODES[i].length()) == COMMAND_CODES[i]) {
+        GCODE_BUFFER_STRING.remove(0, COMMAND_CODES[i].length());
+        code = i;
+        return true;
+      }
+    }
+    GCODE_BUFFER_STRING.remove(0, 1);
+  }
+  return false;
+}
+
+bool Gcode::received_newline() {
+  if (Serial.available()){
+    if (SERIAL_BUFFER_STRING.length() > MAX_SERIAL_BUFFER_LENGTH) {
+      SERIAL_BUFFER_STRING = "";
+    }
+    SERIAL_BUFFER_STRING += Serial.readStringUntil('\n');
+    if (SERIAL_BUFFER_STRING.charAt(SERIAL_BUFFER_STRING.length() - 1) == '\r') {
+      return true;
+    }
+  }
+  return false;
+}
+
+void Gcode::send_ack() {
+  Serial.println("ok");
+  Serial.println("ok");
+}
+
+bool Gcode::read_number(char key) {
+  // NOTE: Has a flaw where if there's 2 `set_temperature` gcodes in the buffer,
+  //       with the first one without a hold time and the next having one,
+  //       then the next hold time will be captured and associated with the
+  //       first gcode
+  int starting_key_index = GCODE_BUFFER_STRING.indexOf(key);
+  if (starting_key_index >= 0) {
```

Comment:
Could we use [strtof](https://en.cppreference.com/w/cpp/string/byte/strtof) instead of this?